### PR TITLE
Addition of token authentication (http and websocket) with Query API and error handling

### PR DESCRIPTION
**Pull request by dannymeloy**
Sorry in advance for the large PR. This is an implementaiton of the web router with a sign-in feature that authenticates using OAuth when pointed at a suitably protected Query API instance (who's endpoints are protected using the class found in: https://github.com/bbc/rd-apmm-python-oauth/blob/master/nmosoauth/resource_server/nmos_security.py). 

I have a branch on the query API you can checkout with protected endpoints (https://github.com/bbc/nmos-query/tree/dannym-oauth).

The feature should:
- Not bother people who aren't pointed at a non-protected query API
- Bother people who are (with appropriate notifications to sign in)
- Detect an nmos-oauth mDNS instance and allow sign in with credentials using the Sign In button
- Authenticate with the Query API by adding `Authorization` headers for HTTP requests
- Add the access token to the query params of the websocket href and authenticate the websockets too
- Allow subsequent sign-ins that trigger a re-render in case of a change of user

### CODE REVIEW

**Comment by PTaylour** [comment]
shall check this apres lunch!

File: src/web-router/dispatchers/security/signin-request.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+export default (username, password) => {
+  let scope = 'is05'
+  let data = {
+    grant_type: 'password',
+    scope: scope,
+    username: username,
+    password: password
+  }
+
+  return getServiceUrl('nmos-auth', 'v1.0', 105)
```
Comment (simonrankine) [inline src/web-router/dispatchers/security/signin-request.js]: Remove magic number
Document behaviour with auth server priority in query string when registry and auth server are on differing priorities in a README

**Comment by simonrankine** [comment]
Reminder to add a ticket to pivitol about switching to authorisation code grant

File: src/ips-nmos-api/src/http/subscription/add-token.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+      if (window.sessionStorage.getItem(TOKEN_KEY)) {
+        let accessToken = window.sessionStorage.getItem(TOKEN_KEY)
+        return accessToken
+      }
+    },
+    addAuthHeaders (headers) {
+      let accessToken = this.fetch()
+      if (accessToken) {
+        let authString = `Bearer ${accessToken}`
+        if (headers) headers['Authorization'] = authString
```
Comment (PTaylour) [inline src/ips-nmos-api/src/http/subscription/add-token.js]: Style point. Generally I don't see many peeps doing single line if statements in js.

Mostly due to these kinda bugs https://medium.com/@jonathanabrams/single-line-if-statements-2565c62ff492

File: src/ips-nmos-api/src/http/subscription/add-token.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
+module.exports = () => {
```
Comment (PTaylour) [inline src/ips-nmos-api/src/http/subscription/add-token.js]: As you're using webpack ya can use the es6 import export system.


```
export const fetch ....
export const addAuthHeaders ....
export const addAuthQuery ....
```

Comment (PTaylour) [reply src/ips-nmos-api/src/http/subscription/add-token.js]: or `export function fetch () { ....`

Comment (dannymeloy) [reply src/ips-nmos-api/src/http/subscription/add-token.js]: I've kept it the same as the other modules in that directory for consistency.. Think Luke uses the CommonJS style for the backend and the ES6 imports style for the app

**Pull request merged**